Complete sn32f2xx support, add functionality

dexter93 opened the pull request:
Supports all sn32f2xx chips
Add options for flash erase, CS re/set and data verification
Simplify and correct the protocol usage

**Automagic** chip identification, no longer VID/PID bound

dexter93 committed:
add 290 support

dexter93 committed:
add 240C support

dexter93 committed:
abstract buffer size

dexter93 committed:
abstract hid reports size from actual payload

dexter93 committed:
rename commands to properly reflect ISP protocol

dexter93 committed:
v2.0.1

abstract functions from the flash function
further compliment the api, introduce support for new chips
actually validate the flash status
new options
fix various bugs

dexter93 committed:
debug: print out verbose fails

dexter93 committed:
debug: try to print error states

this is platform dependent - not implemented in all platforms

dexter93 committed:
correct cs location

dexter93 committed:
add better password checks

and bail if any part fails

dexter93 committed:
print out supported vid pid for known chips

dexter93 committed:
fail if chip version is not supported

dexter93 committed:
fix bad --offset argument reqs

dexter93 committed:
280: preliminary support

dexter93 committed:
erase: correct behavior

we should parse pages, not kb size

dexter93 committed:
add 280 series PID

dexter93 committed:
update buffer handling to send 64b always

dexter93 committed:
rename isp password check

we're actually doing code option table checks here

dexter93 committed:
slight cleanup

dexter93 committed:
v2.0.2 bump

dexter93 committed:
fix command offsets

dexter93 committed:
introduce a mechanism to retry if busy

sometimes we poll too fast. Check if the device is busy and retry

dexter93 committed:
file ops: do NOT keep it open until exit

hopefully this fixes memory corruption

dexter93 committed:
always cleanup on exit

dexter93 committed:
document --noofset

dexter93 committed:
update the docs

dexter93 committed:
standardize logging setup

dexter93 committed:
v2.0.3

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
-{
-    for(int i = 0; i < lenght; i++) data[i] = 0;
+void clear_buffer(unsigned char *data, size_t length) {
+    for (int i = 0; i < length; i++)
+        data[i] = 0;
```
```suggestion
    {
        data[i] = 0;
    }
```

Jpe230 replied:
```suggestion
    for (int i = 0; i < length; i++) {
        data[i] = 0;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+void print_buffer(unsigned char *data, size_t length) {
     printf("Sending Report...\n");
-    for(int i = 0; i < lenght; i++) printf("%02x", data[i]);
+    for (int i = 0; i < length; i++)
+        printf("%02x", data[i]);
```
```suggestion
    {
        printf("%02x", data[i]);
    }
```

Jpe230 replied:
```suggestion
    for (int i = 0; i < length; i++) {
        printf("%02x", data[i]);
    }
```

Jpe230 replied:
```suggestion
    for (int i = 0; i < length; i++) {
        printf("%02x", data[i]);
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
-    {
-        if(oem_reboot && reboot_fail)
-        {
+    if (attempt_no > MAX_ATTEMPTS) return false;
```
```suggestion
    if (attempt_no > MAX_ATTEMPTS)
    {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
-    if(!hid_set_feature(dev,buf, sizeof(buf))) return false;
+    write_buffer_32(buf, command[0]);
+    write_buffer_32(buf + sizeof(uint32_t), command[1]);
+    if (!hid_set_feature(dev, buf, sizeof(buf))) return false;
```
```suggestion
    if (!hid_set_feature(dev, buf, sizeof(buf))) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    write_buffer_16(buf + 4, code_option);
     uint8_t attempt_no = 1;
-    while(!hid_set_feature(dev, buf, 65) && attempt_no <= MAX_ATTEMPTS) // Try {MAX ATTEMPTS} to init flash.
+    while (!hid_set_feature(dev, buf, REPORT_SIZE) && attempt_no <= MAX_ATTEMPTS) // Try {MAX ATTEMPTS} to init flash.
     {
```
```suggestion
    while (!hid_set_feature(dev, buf, REPORT_SIZE) && attempt_no <= MAX_ATTEMPTS) { // Try {MAX ATTEMPTS} to init flash.
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
-        {
+    if (attempt_no > MAX_ATTEMPTS) return false;
+
+    if (!hid_get_feature(dev, buf, CMD_GET_FW_VERSION)) return false;
```
```suggestion
    if (!hid_get_feature(dev, buf, CMD_GET_FW_VERSION)) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+
+    if (!hid_get_feature(dev, buf, CMD_GET_FW_VERSION)) return false;
+    chip = sn32_decode_chip(buf);
+    if (chip == 0) return false;
```
```suggestion
    if (chip == 0) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    chip = sn32_decode_chip(buf);
+    if (chip == 0) return false;
+    cs_level = sn32_get_cs_level(buf);
+    if (!sn32_check_isp_code_option(buf)) return false;
```
```suggestion
    if (!sn32_check_isp_code_option(buf)) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
             fprintf(stderr, "ERROR: Failed to initialize: response cmd is 0x%08x, expected 0x%08x.\n", resp, 0);
-        }
-        else fprintf(stderr, "ERROR: Failed to initialize: response cmd is 0x%08x, expected 0x%08x.\n", resp, CMD_INIT);
+        } else
+            fprintf(stderr, "ERROR: Failed to initialize: response cmd is 0x%08x, expected 0x%08x.\n", resp, CMD_VERIFY(CMD_GET_FW_VERSION));
```
```suggestion
        } else {
            fprintf(stderr, "ERROR: Failed to initialize: response cmd is 0x%08x, expected 0x%08x.\n", resp, CMD_VERIFY(CMD_GET_FW_VERSION));
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    buf[0] = CMD_COMPARE_CODE_OPTION;
+    write_buffer_16(buf + 1, CMD_BASE);
+    write_buffer_16(buf + 4, code_option);
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
```
```suggestion
    if (!hid_set_feature(dev, buf, REPORT_SIZE)) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    write_buffer_16(buf + 1, CMD_BASE);
+    write_buffer_16(buf + 4, code_option);
+    write_buffer_16(buf + 6, CS0); // WARNING THIS SETS CS0
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
+    if (!hid_get_feature(dev, buf, CMD_SET_ENCRYPTION_ALGO)) return false;
```
```suggestion
    if (!hid_set_feature(dev, buf, REPORT_SIZE) || !hid_get_feature(dev, buf, CMD_SET_ENCRYPTION_ALGO)) {
        return false;
    }
```

Jpe230 replied:
```suggestion
    if (!hid_set_feature(dev, buf, REPORT_SIZE) || !hid_get_feature(dev, buf, CMD_SET_ENCRYPTION_ALGO)) {  
        return false;  
    } 
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    buf[0] = CMD_ENABLE_ERASE;
+    write_buffer_16(buf + 1, CMD_BASE);
+    write_buffer_16(buf + 8, USER_ROM_PAGES);
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
+    if (!hid_get_feature(dev, buf, CMD_ENABLE_ERASE)) return false;
```
```suggestion
    if (!hid_set_feature(dev, buf, REPORT_SIZE) || !hid_get_feature(dev, buf, CMD_ENABLE_ERASE)) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    clear_buffer(buf, REPORT_SIZE);
+    buf[0] = CMD_RETURN_USER_MODE;
+    write_buffer_16(buf + 1, CMD_BASE);
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
```
```suggestion
    if (!hid_set_feature(dev, buf, REPORT_SIZE)) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
-    size_t bytes_read = 0;
-    clear_buffer(buf, 65);
-    while ((bytes_read = fread(buf+1, 1, 64, firmware)) > 0)
+    if (chip == SN260 && !flash_jumploader && offset == 0) // Failsafe when flashing a 268 w/o jumploader and offset
     {
```
```suggestion
    if (chip == SN260 && !flash_jumploader && offset == 0) { // Failsafe when flashing a 268 w/o jumploader and offset
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    write_buffer_16(buf + 1, CMD_BASE);
+    write_buffer_32(buf + 4, (uint32_t)offset);
+    write_buffer_32(buf + 8, (uint32_t)(fw_size / REPORT_SIZE));
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
+
+    if (!hid_get_feature(dev, buf, CMD_ENABLE_PROGRAM)) return false;
```
```suggestion
    if (!hid_set_feature(dev, buf, REPORT_SIZE) || !hid_get_feature(dev, buf, CMD_ENABLE_PROGRAM)) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+        if (bytes_read < REPORT_SIZE) {
+            fprintf(stderr, "WARNING: Read %zu bytes, expected %d bytes.\n", bytes_read, REPORT_SIZE);
+        }
+        if (!hid_set_feature(dev, buf, bytes_read)) return false;
```
```suggestion
        if (!hid_set_feature(dev, buf, bytes_read)) {
            return false;
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    // 07) Verify flash complete
+    printf("\n");
+    printf("Verifying flash completion...\n");
+    if (!hid_get_feature(dev, buf, CMD_ENABLE_PROGRAM)) return false;
```
```suggestion
    if (!hid_get_feature(dev, buf, CMD_ENABLE_PROGRAM)) {
        return false;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+int str2buf(void *buffer, char *delim_str, char *string, int buflen, int bufelem_size) {
+    char *s;
+    int   pos = 0;
+    if (string == NULL) return -1;
```
```suggestion
    if (string == NULL) {
        return -1;
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+}
+
+void cleanup(hid_device *handle) {
+    if (handle) hid_close(handle);
```
```suggestion
    if (handle) {
        hid_close(handle);
    }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
         }
         // exit clean after printing
-        if(opt == 'h' || opt == 'V') exit(1);
+        if (opt == 'h' || opt == 'V') exit(1);
```
```suggestion
        if (opt == 'h' || opt == 'V') {
            exit(1);
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
     uint8_t attempt_no = 1;
-    while(handle == NULL && attempt_no <= MAX_ATTEMPTS) // Try {MAX ATTEMPTS} to connect to device.
-    {   
+    while (handle == NULL && attempt_no <= MAX_ATTEMPTS) // Try {MAX ATTEMPTS} to connect to device.
+    {
```
```suggestion
    while (handle == NULL && attempt_no <= MAX_ATTEMPTS) { // Try {MAX ATTEMPTS} to connect to device.
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
-            if(vid == EVISION_VID && !reboot_requested) printf("Warning: eVision VID detected! You probably need to use the reboot option.\n");
-            if(vid == APPLE_VID && !reboot_requested) printf("Warning: Apple VID detected! You probably need to use the reboot option.\n");
+        if (vid != SONIX_VID || (pid != SN248_PID && pid != SN248B_PID && pid != SN248C_PID && pid != SN268_PID && pid != SN289_PID && pid != SN299_PID)) {
+            if (vid == EVISION_VID && !reboot_requested) printf("Warning: eVision VID detected! You probably need to use the reboot option.\n");
```
```suggestion
            if (vid == EVISION_VID && !reboot_requested) {
                printf("Warning: eVision VID detected! You probably need to use the reboot option.\n");
            }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
-            if(vid == APPLE_VID && !reboot_requested) printf("Warning: Apple VID detected! You probably need to use the reboot option.\n");
+        if (vid != SONIX_VID || (pid != SN248_PID && pid != SN248B_PID && pid != SN248C_PID && pid != SN268_PID && pid != SN289_PID && pid != SN299_PID)) {
+            if (vid == EVISION_VID && !reboot_requested) printf("Warning: eVision VID detected! You probably need to use the reboot option.\n");
+            if (vid == APPLE_VID && !reboot_requested) printf("Warning: Apple VID detected! You probably need to use the reboot option.\n");
```
```suggestion
            if (vid == APPLE_VID && !reboot_requested) {
                printf("Warning: Apple VID detected! You probably need to use the reboot option.\n");
            }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
-                    fprintf(stderr, "ERROR: Unsupported sonix bootloader device. Quitting.\n");
-                    exit(1);
-            }
+        if (!ok) error(handle);
```
```suggestion
        if (!ok) {
            error(handle);
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
-            }
+        if (!ok) error(handle);
+        sleep(3);
+        if (chip == SN240 || chip == SN290) ok = protocol_code_option_check(handle); // 240 and 290
```
```suggestion
        if (chip == SN240 || chip == SN290) {
            ok = protocol_code_option_check(handle); // 240 and 290
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
+        if (!ok) error(handle);
+        sleep(3);
+        if (chip == SN240 || chip == SN290) ok = protocol_code_option_check(handle); // 240 and 290
+        if (!ok) error(handle);
```
```suggestion
        if (!ok) {
            error(handle);
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
+        if (chip == SN240 || chip == SN290) ok = protocol_code_option_check(handle); // 240 and 290
+        if (!ok) error(handle);
+        sleep(1);
+        if (cs_level != CS0) ok = protocol_reset_cs(handle);
```
```suggestion
        if (cs_level != CS0) {
            ok = protocol_reset_cs(handle);
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
+        if (!ok) error(handle);
+        sleep(1);
+        if (cs_level != CS0) ok = protocol_reset_cs(handle);
+        if (!ok) error(handle);
```
```suggestion
        if (!ok) {
            error(handle);
        }
```

Jpe230 commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
-            }
+        if (!ok) error(handle);
+        sleep(3);
+        if (chip == SN240 || chip == SN290) ok = protocol_code_option_check(handle); // 240 and 290
+        if (!ok) error(handle);
+        sleep(1);
+        if (cs_level != CS0) ok = protocol_reset_cs(handle);
+        if (!ok) error(handle);
+        sleep(1);
+        if (chip == SN240 || chip == SN290) ok = erase_flash(handle);
+        if (!ok) error(handle);
+        sleep(1);
```
```suggestion
        if (chip == SN240 || chip == SN290) {
            ok = protocol_code_option_check(handle); // 240 and 290
            if (!ok) {
                error(handle);
            }
            sleep(1);
        }
        
        if (cs_level != CS0) {
            ok = protocol_reset_cs(handle);
            if (!ok) {
                error(handle);
            }
            sleep(1);
        }

        if (chip == SN240 || chip == SN290) {
            ok = erase_flash(handle);
            if (!ok) {
                error(handle);
            }
            sleep(1);
        }
```

Jpe230 commented:
Woops, vscode had a fart and each suggestion was sent individually, sorry.... Still, lgtm, just some formatting suggestion....

karliss commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+}
+
+bool hid_get_feature(hid_device *dev, unsigned char *data, uint32_t command) {
+    clear_buffer(data, sizeof(data));
```
Sizeof doesn't work that way. You won't get size of buffer by taking sizeof pointer.

Also any reason for reinventing the wheel and not using memset for this (with or without wrapper).

karliss commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
-{
-    return hid_get_feature_report(dev, data, RESPONSE_LEN + 1);
+uint16_t sn32_get_cs_level(unsigned char *data) {
+    cs_level             = 0;
```
Please choose one either the sn32_get_cs_level changes cs_level directly or returns the value and it's the responsibility of caller to do whatever it wants with it. Currently it's a mix of both.

Also the function name says get_cs_level (which I would expect to be 0, 1, 2, 3), but function returns the protection magic number.

dexter93 replied:
Docs in the Code Option table refer to this as Code Security bits, so cs value reflects it. Perhaps a naming cleanup is needed. Currently not giving the option to the user to select CS mode, since that can result in ISP mode disable and no access to the bootloader - enforced CS0 avoids bricks

karliss commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+
+            // Check the status directly in the data buffer
+            unsigned int status = *((unsigned int *)(data + 4));
+            if (status != CMD_OK) {
```
Might be a good idea to check if first 4 bytes match the command (cmd 0xaa 0x55 0x00). Not just that it was some response with OK bytes, which could be for completely different request. 

I assume that's why you were passing the command to this function.

dexter93 replied:
that's an improvement, left to be implemented later. This gets called right after each command, and they're called in series, so it's safe for now

karliss commented on sonixflasher.c:
```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
-            }
+        if (!ok) error(handle);
+        sleep(3);
+        if (chip == SN240 || chip == SN290) ok = protocol_code_option_check(handle); // 240 and 290
```
Are you sure about these version checks? The checks I saw where for != 268 and != 248b which leaves quite a few more models than just 240 and 290.

dexter93 replied:
no really sure about those, this should be updated indeed

karliss commented on sonixflasher.c:
```diff
@@ -9,382 +9,788 @@
+    write_buffer_16(buf + 6, CS0); // WARNING THIS SETS CS0
+    if (!hid_set_feature(dev, buf, REPORT_SIZE)) return false;
+    if (!hid_get_feature(dev, buf, CMD_SET_ENCRYPTION_ALGO)) return false;
+    clear_buffer(buf, REPORT_SIZE);
```
Useless clearing of local stack buffer which is about to go out of scope. Same thing in next few functions.

dexter93 replied:
Mostly being extra careful with buffer management here while implementing undocumented bits of the protocol

karliss reviewed (commented):
Minor nitpick for the description of PR - this doesn't cover all the SN32F200 chips.  It doesn't take into account SN32f239 and SN32f229. Not suggesting it's necessary to support them, just that the description could be adjusted.

dexter93 commented:
> Minor nitpick for the description of PR - this doesn't cover all the SN32F200 chips. It doesn't take into account SN32f239 and SN32f229. Not suggesting it's necessary to support them, just that the description could be adjusted.

Mostly avoided them because of bad/missing docs. However, looking at it again the protocol is compatible and they're simply a revision of 240
```
index,bl_version,family,revMajor,revMinor
SN32F249
12,0x01 0x20 0x00 0x03
SN32F239
13,0x01 0x20 0x00 0x02
SN32F229
14,0x01 0x20 0x00 0x01
```
That also means the version check atm is buggy, since it will identify 230 and 220 chips as 240 ( issues arising from this are  related to flash size on erase/program). Definitely a nice find, will amend

dexter93 committed:
support sn32f220 and sn32f230

dexter93 committed:
cleanup security level checks

dexter93 committed:
fixup protocol checks for 240b and 260

dexter93 committed:
abstract page erase

dexter93 committed:
refactor cs reset

dexter93 committed:
fixup buffer size handling in hid_get_feature

dexter93 committed:
ensure we're checking the correct reply

dexter93 committed:
add 220 and 230 pids, refactor pid check

dexter93 committed:
add flash erase verification

dexter93 committed:
add flash program checksum to verification

dexter93 committed:
v2.0.4

karliss reviewed (approved):
Looks good enough for me.

With which chips has the new code been actually tested?

dexter93 commented:
Testing being done on specific platform samples
240b works ( covers also 260)
290 works ( covers also 280, 240C)
240 tested up until v2.0.3 - worked fine ( should also cover the now unobtainable 220/230)

HorrorTroll reviewed (approved):
240 series work good on latest commit.

dexter93 committed:
cleanup code option print

dexter93 committed:
speed up procedures

dexter93 committed:
update docs and pid tables

dexter93 committed:
persist while attempting to reboot to bootloader

dexter93 committed:
add sonix reboot option

dexter93 committed:
exit if a program operation is already pending

If flashing gets interrupted while the device is receiving the
firmware, it will stay in program mode. Inform the user about it

dexter93 committed:
v2.0.5
